docs/design/proposals/native-binding/native-binding.md

```diff
@@ -0,0 +1,150 @@
+
+This design aims to address two separate use cases:
+
+1. An application team manages their own ACK resources and K8s deployments. They wish to dynamically link the status of an ACK resource to the environment variables of their deployment.
+2. An infrastructure team manages the ACK resources on behalf of application teams. The application teams do not have read/write access to the ACK resource directly, but require access to the fields from them in order to deploy their application.
+
+## Requirements
+
+1. Export any number of Spec and/or Status values to application consumable formats (ConfigMap and Secrets)
+2. Export to any namespace (for cluster-scoped controllers only)
+3. The exported value should stay in sync as the resource changes
+4. Resources should not need to be updated to add/remove bindings
+
+## Existing Solutions
+
+### Redhat Service Binding Operator
+
+>“[Redhat Service Binding Operator] enables developers to connect their application to backing services with a consistent and predictable experience“ - https://github.com/redhat-developer/service-binding-operator
+
+
+The [Redhat Service Binding Operator](https://github.com/redhat-developer/service-binding-operator) (herein called the SBO) uses annotations present on CRs and resources to import field values from other resources. The main purpose of SBO is to standardise the system for injecting secret information from database services into application deployment environment variables. The SBO can automatically detect and bind to services only from a limited number of [supported operators](https://github.com/redhat-developer/service-binding-operator#known-bindable-operators). 
+
+SBO is not suitable for this design as it can only support a limited number of fields based on pre-configured database bindings. That is, it can handle the case where an application developer wants to inject a PostgreSQL database connection string into a deployment, but little more than that. SBO does not support exporting to configmap or secrets directly.
+
+## Proposed Solution
+
+### ACK Binding CRD
+
+ACK uses a common CRD to support resource adoption, implemented through a separate (common) reconciler loop in each controller. Common CRDs allow customers to define controller configuration for tasks that do not fit within the spec of existing CRDs or that are too complex to fit into the annotations. 
+
+This solution would introduce a new CRD solely for the purpose of declaring field exports. The CRD spec would contain an array of structs, each containing:
+
+* The source path, containing:
+    * Resource API group, kind, namespace, name
+* Spec/Status field path
+* An output paths, containing:
+    * Type: ConfigMap or Secret
+    * Namespace and name
+    * (Optionally) Output structure
```

Comment:
> ready made credential available without additional configuration 

How would you envision this working? AFAIK credentials are typically ingested through environment variables or volumes mounted directly to the pod. This solution aimed to make it easy to mount them as environment variables, by referencing a generated configmap/secret. 

Are you suggesting that we provide preformatted connection URLs as the output?

Reply:
@RedbackThomson 

> How would you envision this working? AFAIK credentials are typically ingested through environment variables or volumes mounted directly to the pod. This solution aimed to make it easy to mount them as environment variables, by referencing a generated configmap/secret.

We're on the same page here. My thought is having the Secret/ConfigMap automatically generated by a specific controller when we already know the format of the credentials.

For example, let's use ACK for RDS. We know the [username / password](https://aws-controllers-k8s.github.io/community/docs/tutorials/rds-example/#postgresql) because that is part of custom resource creation. We will get the [host/port](https://aws-controllers-k8s.github.io/community/docs/tutorials/rds-example/#connect-to-database-instances) during the RDS provisioning process.

From a RDS perspective, these are "well-known" attributes, and it would simplify things for users if we are able to create the initial Secret with the credentials for them.

Perhaps in this case, the controller could generate an ACK Binding CRD?

Reply:
> My thought is having the Secret/ConfigMap automatically generated by a specific controller when we already know the format of the credentials

This scares me. I wouldn't want to dump database credentials into K8s without someone explicitly allowing it. I'm thinking of a case where someone has permissions to decrypt secrets within a namespace and can now suddenly see the database password of every RDS database in that namespace. 

I would like to eventually publish some templates for creating an RDS database, including all VPC resources, as a Helm chart. I think it would be great to include this exported field as a `exportCredentials: true/false` value in that chart, though.